### PR TITLE
feat: use same issuer for monthly and annual products

### PR DESCRIPTION
**Pull request by pavelbrm**
### Summary

This PR changes the way SKU values are used when interacting with CBP in such a way as to make sure that each Premium product pair (monthly and annual) uses the same issuer.

Currently, Premium products have SKU values in the format of `brave-[name]-premium`. Annual products will temporarily have SKUs like `brave-[name]-premium-year`. Although eventually variants of the same product will use the same SKU, until that happened we need a way to make sure the same issuer is used for signing and verifying credentials.

This is accomplished by trimming the `-year` suffix from SKU values in places where they are sent to CBP.

Additionally, this PR offers some refactoring and improvements to the verify credential endpoints.

### Type of Change

- [x] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [x] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before Requesting Review

- [x] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [x] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [x] Is there a clear title that explains what the PR does?
- [x] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security and/or privacy review if needed
- [x] Have you performed a self review of this PR?

### CODE REVIEW

**Comment by pavelbrm** [comment]
> [[puLL-Merge](https://github.com/brave/pull-merge)] - [brave-intl/bat-go@2654](https://github.com/brave-intl/bat-go/pull/2654)
> 

If only robot you knew how out of place and wrong your comments are.

> ### Possible Issues
> 1. The removal of `inputs.DecodeAndValidateReader` in favor of manual parsing and validation might introduce edge cases that were previously handled by the library.

No, it does not. The removed code was a convoluted piece, and the new code performs the same steps, but much cleaner. Plus. the new code has got tests.

> 2. The change in SKU handling (trimming "-year" suffix) might affect existing data or integrations that rely on the exact SKU format.

This is the intended change and the whole point of this PR. 🤦‍♀️ 


> ### Security Hotspots
> 1. The new parsing functions (`parseVerifyCredRequestV2` and `parseVerifyCredeOpaque`) handle base64 decoding and JSON unmarshaling of user input. Ensure proper input sanitization and error handling to prevent potential injection attacks or denial of service due to malformed input.

If this LLM had a little clue about what's done in this PR, it would have understood that the new functions do the same thing, but better, and are covered with tests, unlike the removed piece of convoluted logic.

> 2. The `validateVerifyCredRequestV2` function now performs manual validation. Ensure that all necessary validations are included to maintain the security level of the previous implementation.

Again, wrong. No, it does not perform manual validation. It uses a better and less error prone validator.

File: services/skus/controllers_noint_test.go [hunk 1]
```diff
@@ -222,3 +223,151 @@ func TestHandleReceiptErr(t *testing.T) {
...
+		exp   tcExpected
+	}
+
+	tests := []testCase{
+		{
+			name:  "error_malformed_payload",
+			given: []byte(`nonsense`),
+			exp: tcExpected{
+				errFn: func(tt must.TestingT, err error, i ...interface{}) {
+					must.Equal(tt, true, err != nil)
```
Comment (clD11) [inline services/skus/controllers_noint_test.go:245]: question: would using `must.NotNil(tt, err)` not be more concise and improve readability?

File: services/skus/controllers_noint_test.go [hunk 2]
```diff
@@ -222,3 +223,151 @@ func TestHandleReceiptErr(t *testing.T) {
...
+				},
+			},
+		},
+
+		{
+			name:  "error_malformed_credential",
+			given: []byte(`{"sku":"sku","merchantId":"merchantId"}`),
+			exp: tcExpected{
+				errFn: func(tt must.TestingT, err error, i ...interface{}) {
+					must.Equal(tt, true, err != nil)
```
Comment (clD11) [inline services/skus/controllers_noint_test.go:255]: Same as above, would using `must.NotNil(tt, err)` not be more concise?

File: services/skus/controllers_noint_test.go [hunk 3]
```diff
@@ -222,3 +223,151 @@ func TestHandleReceiptErr(t *testing.T) {
...
+					SKU:        "sku",
+					MerchantID: "merchantId",
+					Credential: "eyJ0eXBlIjoidGltZS1saW1pdGVkLXYyIiwicHJlc2VudGF0aW9uIjoiVG1GMGRYSmxJR0ZpYUc5eWN5QmhJSFpoWTNWMWJTNEsifQo=",
+					CredentialOpaque: &model.VerifyCredentialOpaque{
+						Type:         "time-limited-v2",
+						Presentation: "TmF0dXJlIGFiaG9ycyBhIHZhY3V1bS4K",
+					},
+				},
+				errFn: func(tt must.TestingT, err error, i ...interface{}) {
+					must.Equal(tt, true, err == nil)
```
Comment (clD11) [inline services/skus/controllers_noint_test.go:274]: question: would using `must.Nil(tt, err)` be more concise?

File: services/skus/controllers_noint_test.go [hunk 4]
```diff
@@ -222,3 +223,151 @@ func TestHandleReceiptErr(t *testing.T) {
...
+			},
+		},
+	}
+
+	for i := range tests {
+		tc := tests[i]
+
+		t.Run(tc.name, func(t *testing.T) {
+			actual, err := parseVerifyCredRequestV2(tc.given)
+			tc.exp.errFn(t, err)
```
Comment (clD11) [inline services/skus/controllers_noint_test.go:285]: nit: naming the function `mustErr` opposed to `errFn` would give more indication that this will not continue and will perform an assertion.

File: services/skus/controllers.go [hunk 5]
```diff
@@ -943,54 +947,79 @@ func MerchantTransactions(service *Service) handlers.AppHandler {
...
 			return handlers.WrapError(err, "Error in request body", http.StatusBadRequest)
 		}
 
-		appErr := service.verifyCredential(ctx, req, w)
-		if appErr != nil {
-			l.Error().Err(appErr).Msg("failed to verify credential")
+		if err := validateVerifyCredRequestV2(valid, req); err != nil {
+			verrs, ok := collectValidationErrors(err)
+			if !ok {
+				return handlers.ValidationError("request", map[string]interface{}{"request-body": err.Error()})
```
Comment (clD11) [inline services/skus/controllers.go:973]: question: will this change format of the response to the client?

**Comment by clD11** [comment]
The branch conflicts will also need to be be resolved too

File: services/skus/controllers.go [hunk 6]
```diff
@@ -943,54 +947,79 @@ func MerchantTransactions(service *Service) handlers.AppHandler {
...
+		data, err := io.ReadAll(io.LimitReader(r.Body, reqBodyLimit10MB))
+		if err != nil {
+			lg.Warn().Err(err).Msg("failed to read body")
+
+			return handlers.WrapError(err, "Error in request body", http.StatusBadRequest)
+		}
+
+		req, err := parseVerifyCredRequestV2(data)
+		if err != nil {
+			lg.Warn().Err(err).Msg("failed to deserialize request")
```
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/controllers.go:964]: This text is used in at least 5 other places I've changed, and those passed previous reviews. You reviewed ]the fist PR](https://github.com/brave-intl/bat-go/pull/1916#discussion_r1287235021) that used this text, and since then there had been no questions.

This is consistent, and I don't see an issue with this. While the name of the function contains `parse`, the process involves a bit more than that, and it seemed appropriate to generalise and refer to it as deserialisation.

Thoughts?

File: services/skus/controllers_noint_test.go (hunk 1 above)
Comment (pavelbrm) [reply services/skus/controllers_noint_test.go:245]: Personally, I do not like tying ourselves to the API of a dependency more and more. This is syntactic sugar, and I don't believe this sugar is necessary. When using a dependency, I prefer using as less as possible.

I've done it in other places before, and those have been passed without a question in past reviews.

My preference would be to stick to the basics unless there is a big win. Using the assertion library is a win, but spreading all of its various methods over the code base does not look appealing.

Happy to discuss it further.

File: services/skus/controllers_noint_test.go (hunk 2 above)
Comment (pavelbrm) [reply services/skus/controllers_noint_test.go:255]: Same reasoning as [here](https://github.com/brave-intl/bat-go/pull/2654#discussion_r1756217159).

File: services/skus/controllers_noint_test.go (hunk 3 above)
Comment (pavelbrm) [reply services/skus/controllers_noint_test.go:274]: Same reasoning as [here](https://github.com/brave-intl/bat-go/pull/2654#discussion_r1756217159).

File: services/skus/controllers_noint_test.go (hunk 4 above)
Comment (pavelbrm) [reply services/skus/controllers_noint_test.go]: I thought about this, and liked `errFn` more, and chose it. However, updated according to your suggestion. Thanks.

File: services/skus/controllers.go (hunk 5 above)
Comment (pavelbrm) [reply services/skus/controllers.go]: It will, I guess. Thought about this, and was choosing between a better representation and preservation of the format. Went with the improvement. However, you're right – if this error is meaningful to and interpreted by the client, this might create an issue. Restored the previous format as much as it was possible.

File: services/skus/controllers_noint_test.go (hunk 1 above)
Comment (clD11) [reply services/skus/controllers_noint_test.go:245]: This library and its API are used everywhere in the code base and across almost all of our repos including `must.NotNil(tt, err)` so its not clear how adding this convention for asserting on errors is going to alleviate that issue. Would this reasoning also apply to `must.NoError(t, err)` which is prevalent in almost every test we have?

Syntactic sugar is there to make code more readable. Working in the negative and passing in the condition, like above requires more 'mental gymnastics' than simply using the provided API and passing the thing you are asserting on e.g. `must.NotNil(tt, err)`.

**Comment by clD11** [comment]
Now the [blocking comment](https://github.com/brave-intl/bat-go/pull/2654#discussion_r1755670709) and conflicts have been resolved, happy to approve.

Left a few non blocking responses to your comments.

**Review by clD11 — APPROVED** [review]

File: services/skus/controllers.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/controllers.go:964]: Yes, the link is correct – it shows the place where I had this text first.

> Arguably deserilaztion is more specific than parsing so to generalise it might be better the other way around.

Yes, agreed. I imprecisely expressed what I was hoping to say. Instead of `generalise`, it should have been `specify`. Because what's happening in the cases I used this message is deserialisation (i.e. parsing and constructing an object).

However, since the functions contain `parse` in their names it seems appropriate to log with `parse` as well.

Let me update it.

File: services/skus/controllers_noint_test.go (hunk 1 above)
Comment (clD11) [reply services/skus/controllers_noint_test.go:245]: That's a good point maybe we should updated those tests as well.

This approach was added to the code base along with many other changes to testing which we adopted however, over time this particular feature hasn't proven to add much value hence raising the question now. 

For example, take [this](https://github.com/brave-intl/bat-go/blob/35f7182b31c0b6c7625f4c3359f06df59f4c40d1/services/skus/storage/repository/issuer_test.go#L256) test. There are many steps where we don't want an error before moving on and it is considerably more concise and clear to the reader to simply use the API -

```
must.ErrorIs(t, err, tc.exp.err)

must.NoError(t, err)

```
opposed to whats being advocated above

```
must.Equal(t, true, errors.Is(err, tc.exp.err))

must.Equal(tt, true, err != nil)
```

Comment (pavelbrm) [reply services/skus/controllers_noint_test.go:245]: Like I said, I agree with your reasoning, and also have my view on this subject. The choice in this specific PR has been motivated mainly by consistency with what I've done so far.

In one of the future PRs, I will update this and other mentioned tests in accordance with your suggestion as it would allow us to minimise the cognitive load while working with our systems (that already have a fair bit of accidental complexity in them).
